File: part-02/src/main/java/eu/luminis/workshop/smallsteps/logic/domainservice/LegoStock.java
package eu.luminis.workshop.smallsteps.logic.domainservice;

import eu.luminis.workshop.smallsteps.logic.domainmodel.valueobjects.LegoBuilderId;
import eu.luminis.workshop.smallsteps.logic.domainmodel.valueobjects.LegoSetNumber;
import eu.luminis.workshop.smallsteps.logic.domainservice.auth.AuthProvider;
import eu.luminis.workshop.smallsteps.logic.domainservice.catalog.LegoPartCatalog;
import eu.luminis.workshop.smallsteps.logic.domainservice.state.IncompleteReturn;
import eu.luminis.workshop.smallsteps.logic.domainservice.state.LegoBox;
import eu.luminis.workshop.smallsteps.logic.domainservice.state.LegoSetForBuilder;
import eu.luminis.workshop.smallsteps.logic.domainservice.state.StockState;

import java.util.*;

public class LegoStock {
    private final AuthProvider authProvider;
    private final LegoPartCatalog legoPartCatalog;
    private final StockState stockState;

    public LegoStock(AuthProvider authProvider, LegoPartCatalog legoPartCatalog, StockState stockState) {
        this.authProvider = authProvider;
        this.legoPartCatalog = legoPartCatalog;
        this.stockState = stockState;
    }

    public StockState reserve(LegoSetNumber legoSetNumber) {
        LegoBuilderId forBuilder = authProvider.currentAuthentication().requiredBuilderId();

        Integer inStock = stockState.getAvailableStock().getOrDefault(legoSetNumber, 0);
        require(inStock > 0, String.format("LegoSet %d is not in stock", legoSetNumber.getNumber()));

        return stockState.copy(
                reduceStockByOne(legoSetNumber, inStock),
                addReservation(legoSetNumber, forBuilder),
                null,
                null,
                null);
    }

    public StockState ship(LegoSetNumber legoSetNumber, LegoBuilderId forBuilder) {
        authProvider.currentAuthentication().mustBeStore();
        check(stockState.getReserved().contains(new LegoSetForBuilder(legoSetNumber, forBuilder)),
                String.format("Legoset %d has not been reserved by builder %s",
                        legoSetNumber.getNumber(), forBuilder.toString()));


        return stockState.copy(
                null,
                removeReservation(legoSetNumber, forBuilder),
                addAtBuilder(legoSetNumber, forBuilder),
                null,
                null);
    }

    public StockState returnComplete(LegoSetNumber legoSetNumber) {
        LegoBuilderId byBuilder = authProvider.currentAuthentication().requiredBuilderId();

        check(stockState.getAtBuilder().contains(new LegoSetForBuilder(legoSetNumber, byBuilder)),
                String.format("Cannot return %s as builder %s does not have it.",
                        legoSetNumber.getNumber(), byBuilder.toString()));

        return stockState.copy(
                null,
                null,
                null,
                null,
                null);
    }

    public StockState returnInComplete(LegoBox legoBox) {
        LegoBuilderId byBuilder = authProvider.currentAuthentication().requiredBuilderId();
        check(stockState.getAtBuilder().contains(new LegoSetForBuilder(legoBox.getLegoSetNumber(), byBuilder)),
                String.format("Cannot return %s as builder %s does not have it.",
                        legoBox.getLegoSetNumber().getNumber(), byBuilder.toString()));

        this.requireValidMissingLegoParts(legoBox);

        return stockState.copy(
                null,
                null,
                null,
                addIncompleteStock(legoBox),
                addIncompleteReturnHistory(legoBox));
    }

    private List<IncompleteReturn> addIncompleteReturnHistory(LegoBox legoBox) {
        return new ArrayList<>() {
            {
                addAll(stockState.getIncompleteReturnHistory());
                add(new IncompleteReturn(legoBox.getLegoSetNumber(), legoBox.getMissingParts()));
            }
        };
    }

    private List<LegoBox> addIncompleteStock(LegoBox legoBox) {
        return new ArrayList<>() {
            {
                addAll(stockState.getIncompleteStock());
                add(legoBox);
            }
        };
    }

    private Set<LegoSetForBuilder> addReservation(LegoSetNumber legoSetNumber, LegoBuilderId forBuilder) {
        return new HashSet<>() {
            {
                addAll(stockState.getReserved());
                add(new LegoSetForBuilder(legoSetNumber, forBuilder));
            }
        };
    }

    private Map<LegoSetNumber, Integer> reduceStockByOne(LegoSetNumber legoSetNumber, Integer inStock) {
        return new HashMap<>() {
            {
                putAll(stockState.getAvailableStock());
                put(legoSetNumber, inStock - 1);
            }
        };
    }

    private List<LegoSetForBuilder> addAtBuilder(LegoSetNumber legoSetNumber, LegoBuilderId forBuilder) {
        return new ArrayList<>() {
            {
                addAll(stockState.getAtBuilder());
                add(new LegoSetForBuilder(legoSetNumber, forBuilder));
            }
        };
    }

    private Set<LegoSetForBuilder> removeReservation(LegoSetNumber legoSetNumber, LegoBuilderId forBuilder) {
        return new HashSet<>() {
            {
                addAll(stockState.getReserved());
                remove(new LegoSetForBuilder(legoSetNumber, forBuilder));
            }
        };
    }

    private void requireValidMissingLegoParts(LegoBox legoBox) {
        require(!legoBox.getMissingParts().isEmpty(), "No missing parts specified");
        legoBox.getMissingParts().forEach((partNumber, numberOfParts) -> {
            require(numberOfParts > 0, "You must specify how many parts are missing");
        });

        Map<String, Integer> expectedParts = legoPartCatalog.allPartsForLegoSet(legoBox.getLegoSetNumber());
        require(!expectedParts.isEmpty(),
                String.format("Unable to determine parts for lego set %s", legoBox.getLegoSetNumber()));

        legoBox.getMissingParts().forEach((partNumber, numberOfParts) -> {
            require(expectedParts.containsKey(partNumber),
                    String.format("Some parts reported missing don't belong to lego set %s",
                            legoBox.getLegoSetNumber()));

            require(expectedParts.getOrDefault(partNumber, 0) >= numberOfParts,
                    String.format("Too many parts reported missing, the original lego set %s did not contain that " +
                            "many of this part %s", legoBox.getLegoSetNumber().getNumber(), partNumber));
        });
    }

    private static void require(boolean check, String message) {
        if (!check) {
            throw new IllegalArgumentException(message);
        }
    }

    private static void check(boolean check, String message) {
        if (!check) {
            throw new IllegalStateException(message);
        }
    }
}

File: part-02/src/main/kotlin/eu/luminis/workshop/smallsteps/logic/domainService/LegoStock.kt
package eu.luminis.workshop.smallsteps.logic.domainService

import eu.luminis.workshop.smallsteps.logic.domainModel.valueObjects.LegoBuilderId
import eu.luminis.workshop.smallsteps.logic.domainModel.valueObjects.LegoSetNumber
import eu.luminis.workshop.smallsteps.logic.domainService.auth.AuthProvider
import eu.luminis.workshop.smallsteps.logic.domainService.catalog.LegoPartCatalog
import eu.luminis.workshop.smallsteps.logic.domainService.state.IncompleteReturn
import eu.luminis.workshop.smallsteps.logic.domainService.state.LegoBox
import eu.luminis.workshop.smallsteps.logic.domainService.state.StockState

data class LegoStock(
    private val authProvider: AuthProvider,
    private val legoPartCatalog: LegoPartCatalog,
    private val state: StockState,
) {
    fun reserve(legoSet: LegoSetNumber): StockState {
        val forBuilder = authProvider.currentAuthentication.requiredBuilderId()

        val inStock = state.availableStock.getOrDefault(legoSet, 0)
        require(inStock > 0) { "LegoSet $legoSet is not in stock" }

        return state.copy(
            availableStock = state.availableStock.mapValues(reduceStockByOne(legoSet)),
            reserved = state.reserved + (legoSet to forBuilder)
        )
    }

    private fun reduceStockByOne(legoSet: LegoSetNumber) = { entry: Map.Entry<LegoSetNumber, Int> ->
        if (entry.key == legoSet) entry.value - 1
        else entry.value
    }

    fun ship(legoSet: LegoSetNumber, toBuilder: LegoBuilderId): StockState {
        authProvider.currentAuthentication.mustBeStore()
        check(state.reserved.contains(legoSet to toBuilder)) { "Legoset $legoSet has not been reserved by builder $toBuilder" }

        return state.copy(
            reserved = state.reserved - (legoSet to toBuilder),
            atBuilder = state.atBuilder + (legoSet to toBuilder),
        )
    }

    fun returningComplete(legoSet: LegoSetNumber): StockState {
        val byBuilder = authProvider.currentAuthentication.requiredBuilderId()
        check(state.atBuilder.contains(legoSet to byBuilder))

        return state
    }

    fun returningInComplete(legoBox: LegoBox): StockState {
        val byBuilder = authProvider.currentAuthentication.requiredBuilderId()
        check(state.atBuilder.contains(legoBox.legoSet to byBuilder))
        legoBox.requireValidMissingLegoParts()

        return state.copy(
            incompleteStock = state.incompleteStock + legoBox,
            incompleteReturnHistory = state.incompleteReturnHistory + IncompleteReturn(
                legoBox.legoSet,
                legoBox.missingParts
            ),
        )
    }

    private fun LegoBox.requireValidMissingLegoParts() {
        require(this.missingParts.isNotEmpty()) { "No missing parts specified" }
        require(this.missingParts.all { it.value > 0 }) { "You must specify how many parts are missing" }

        val expectedParts = legoPartCatalog.allPartsForLegoSet(legoSet)
            ?: throw IllegalStateException("Unable to determine parts for lego set $legoSet")
        require(this.missingParts.all { expectedParts.containsKey(it.key) }) {
            "Some parts reported missing don't belong to lego set $legoSet"
        }
        require(this.missingParts.all { expectedParts.getOrDefault(it.key, 0) >= it.value }) {
            "Too many parts reported missing, the original lego set $legoSet did not contain that many of this part"
        }
    }
}